master/internal/hpimportance/hpimportance.go

```diff
@@ -1,15 +1,154 @@
+		}
+	}
+	st := "@attribute numBatches numeric\n"
+	f.WriteString(st)
+
+	// Now we have to add the data to the file
+	f.WriteString("\n\n@data\n")
+
+	// First, get batch ids unless better way to get them in go?
+	var batches []int
+	for k := range data {
+		batches = append(batches, k)
+	}
+	sort.Sort(sort.Reverse(sort.IntSlice(batches)))
+
+	totalNumTrials := 0
+	maxNumBatches := batches[0]
+	for _, batchId := range batches {
+		if batchId <= maxNumBatches/maxDiffCompBatch {
+			break
+		}
+		if totalNumTrials > minNumberTrials && batchId <= maxNumBatches/idealBatchDiff {
+			break
+		}
+		batchIdStr := fmt.Sprintf("%v", batchId)
+		for _, trial := range data[batchId] {
+			st := ""
+			metric := fmt.Sprintf("%v", trial.Metric)
+			st = st + metric
+			hparamsVals := trial.Hparams
+			for _, hp := range hpsOrder {
+				s := fmt.Sprintf("%v", hparamsVals[hp])
+				st = st + "," + s
+			}
+
+			st = st + "," + batchIdStr + "\n"
+			f.WriteString(st)
+
+			totalNumTrials = totalNumTrials + 1
+		}
+
+	}
+
+	f.Close()
+	return totalNumTrials
+}
+
+func computeHPImportance(data map[int][]model.HPImportanceTrialData, experimentConfig *model.ExperimentConfig, masterConfig HPImportanceConfig, growforest string, workingDir string) (map[string]float64, error) {
+
+	totalNumTrials := createDataFile(data, experimentConfig, workingDir)
+
+	nCores := strconv.FormatInt(int64(masterConfig.CoresPerWorker), 10)
+	maxNumTrees := int64(masterConfig.MaxTrees)
+
+	// random may be smaller because only 50 trials are ran
+	// where I'm not gonna calculate the random forest
+	// TODO: Determine best way to handle small amount of trials
+	if totalNumTrials < minNumberTrials{
+		return errors.New("Not enough trials for HP importance.")
+	}
+
+	// For version one, we do half the number of trials up to 300
+	// This may be overdoing and running too long for minimal performance inc.
+	// The paper uses 2 values replications and ntrees while the library,
+	// only uses --ace which is replications.
+	// TODO: use the bar chart from hp viz to improve hp importance
+	numTrees := int64(totalNumTrials / nReplications)
+
+	if numTrees > maxNumTrees {
+		numTrees = maxNumTrees
+	}
+	strNumTrees := strconv.FormatInt(numTrees, 10)
+
+	// Call Random Forest
+	_, err := exec.Command(growforest, "-train", workingDir+"/data.arff", "-target=metric", "-nCores", nCores, "-ace", strNumTrees, "-importance", workingDir+"/importance.txt", "-rfpred", workingDir+"rface.sf").Output()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Read and Return Output
+	file, err := os.Open(workingDir + "/importance.txt")
+
 	output := make(map[string]float64)
-	for key := range data[0].Hparams {
-		output[key] = 0.0
+	r := csv.NewReader(file)
+	r.Comma = '\t'
+	for {
+		record, err := r.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		if record[1] == "metric" {
+			continue
+		}
+		output[record[1]], _ = strconv.ParseFloat(record[2], 32) // 32 or 64 bit?
```

Comment:
This all feels really brittle; is there really no good way to consume growforest as a lib? I peeked and the answer looked grim lol. If not can we break this into a func called like "parseGrowforestOutput` with a link to the docs that describe the format being parsed?

Reply:
Consuming it as a lib would be nice, although I like having it as a separate process from a resource isolation standpoint. There's a bunch of logic we would need to duplicate in their CLI in order to use it as a lib without modifying their lib a bunch, but the big blocker was that the ability to restrict this to a certain number of CPU cores is done with GOMAXTHREADS and that affects the whole process.